netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -105,6 +132,49 @@ public void onGoAwayReceived(int lastStreamId, long errorCode, ByteBuf debugData
     });
   }
 
+  /**
+   * Sends a ping to the server.
+   */
+  void ping(final PingCallback callback, Executor executor) {
+    checkState(ctx != null);
+    Ping newPing = null;
+    Stopwatch stopwatch = null;
+    ByteBuf buffer = null;
+    while (true) {
+      if (goAwayStatus != null) {
+        Ping.failed(callback, executor, getPingFailure());
+        return;
+      }
+      // we only allow one outstanding ping at a time, so just add the callback to
+      // any outstanding operation
+      Ping p = ping.get();
+      if (p != null) {
+        p.addCallback(callback, executor);
+        return;
+      }
+      if (newPing == null) {
+        // initialize ping payload
+        byte[] bytes = new byte[8];
+        random.nextBytes(bytes);
+        buffer = Unpooled.wrappedBuffer(bytes);
+        stopwatch = Stopwatch.createUnstarted(ticker);
+        newPing = new Ping(bytes, stopwatch);
+      }
+      if (ping.compareAndSet(null, newPing)) {
+        stopwatch.start();
+        frameWriter().writePing(ctx, false, buffer, ctx.newPromise());
```

Comment:
I redid this. I realized that calling `writePing` from an arbitrary app thread could mistakenly break up the two underlying writes (frame header and then payload), interleaving them with other writes.

I changed it to use `ctx.write(Object)` and followed the command pattern already in place for other such writes. That should end up issuing the writes from a `ChannelHandlerInvoker`, which should be correct for expected thread-safety. I'm hoping you or @ejona86 have enough Netty-fu to help me make sure I'm doing it correctly.

In any event, the new code that writes the ping frame now also does a flush. Digging through everything else, it looked like you were right and that it is necessary.